src/main/java/frc/robot/subsystems/DrivetrainSubsystem.java

```diff
@@ -174,10 +178,35 @@ public void resetPose(Pose2d pose) {
         }
 
         private void updatePose() {
+                if(Robot.isSimulation()) {
+                        positionMeters[0] += states[0].speedMetersPerSecond * (Timer.getFPGATimestamp() - lastUpdateTime);
+                        positionMeters[1] += states[1].speedMetersPerSecond * (Timer.getFPGATimestamp() - lastUpdateTime);
+                        positionMeters[2] += states[2].speedMetersPerSecond * (Timer.getFPGATimestamp() - lastUpdateTime);
+                        positionMeters[3] += states[3].speedMetersPerSecond * (Timer.getFPGATimestamp() - lastUpdateTime);
+                        simRotation += kinematics.toChassisSpeeds(states).omegaRadiansPerSecond * (Timer.getFPGATimestamp() - lastUpdateTime);
+
+                        if(DriverStation.getAlliance() == Alliance.Red && robotPose != null){
+                                Pose2d redOrigin = new Pose2d(
+                                        new Translation2d(Constants.RED_ORIGIN_POS_X_METERS, Constants.RED_ORIGIN_POS_Y_METERS),
+                                        Rotation2d.fromDegrees(Constants.RED_ORIGIN_ROTATION_DEG)
+                                );
+                                robotPose = redOrigin.transformBy(new Transform2d(robotPose.getTranslation(), robotPose.getRotation()));
+                        }
```

Comment:
This block of code is independent of whether we are in simulation mode. it can be outside and would work either way.

Best place for it is after we set the robotPose below. however it's important that we don't modify the actual pose but just the pose that gets displayed on the Field2d widget. That is why the code snippet I shared referred to it as displayedPose